### PR TITLE
Monstermos - An atmos system that's better

### PR DESCRIPTION
**Pull request by monster860**
:cl: monster860
add: Air spreads much more quickly now - If there is a pressure difference, it spreads instantly. This means that if you break a window in a hallway, all the air gets sucked out immediately
add: Air vents are now buffed - A single air vent can refill a decently sized room in about a minute now. Likewise, a scrubber can empty a room pretty quickly now.
add: Air vents now have a space shutoff function - if it detects that it's venting is being immediately sucked into space, it will shut off for 20 seconds - during these 20 seconds it will blink yellow to indicate that it is in space shutoff mode.
add: Firelocks now automatically close to prevent air from escaping into space.
add: Pairs of firelocks now automatically airlock to prevent air from escaping when going through
add: Firelocks can now be opened by bumping, if it determines not much air will escape and no fire alarm is triggered
tweak: Distro pumps start out maxed
add: There is now a delay along with a warning message when trying to open a firelock with a pressure difference.
/:cl:

https://streamable.com/xrez1

Things that need done before fully merge ready:

- [x] Make it work on lavaland
- [x] Figure out the inner moles system so it actually works (for now gases may spread farther than they're supposed to)
- [ ] ~~Make airlocks cycle at some point maybe~~ i'll do it in a later PR
- [x] Put firelocks in shuttles I guess

### CODE REVIEW

**Comment by boodaliboo** [comment]
I remain neutral on this a test merge is very much required until we can see how this effects normal rounds

**Comment by ghost** [comment]
This is a great idea. For ages, atmos has been plagued by the computational limitations of LINDA and the creative limitations of ZAS. This serves as the perfect balance between the two and should definitely be tried out when it is finished. Balancing the damage can happen later.

**Comment by MegaEmpirical** [comment]
I'd suggest reducing the timer on opening airlocks. Not uncommon to be trying to flee for your life.

Any changes to how quickly they close automatically after opening? They usually close damn quick and with the push back from the air it'll be near impossible to get through in time.

**Comment by monster860** [comment]
Right now the firelocks close in 3 seconds.

**Comment by Lostlarry** [comment]
What about some making some machine (or modify air alarm) to notify engi over radio in high priority areas

**Comment by MegaEmpirical** [comment]
Pre-rebase we could see alerts on our PDAs for engineering and atmos would even get a message and beep for fires.

I'd love to have this back.

**Comment by Hopekz** [comment]
FYI:
It's not as extreme as it was during monster's initial videos.
Here is me going through the xeno spacewalk, breaking arrivals and breaking escape
(all tests done on a complete reset to not to affect the pressures in any way)
https://imgur.com/a/sjx3Ive

+1

**Comment by MegaEmpirical** [comment]
Will you be adjusting air locks to have the little fans so that people are not jettisoned when using them?

**Comment by monster860** [comment]
Fans are pretty snowflakey, I'd rather have proper cycling airlocks.

**Comment by MegaEmpirical** [comment]
As long as one or the other is done before merging.

**Comment by swissloaf** [comment]
image

**Comment by 00aquir** [comment]
very bad, if there is pressure difference near mining shuttle doors, the miners seemingly often get thrown near the rooms connecting to shuttle, when I was miner, it immediately injured me as I was coming back, as it threw me and my ore box, it can also block the mining dock external with the crate there, as pulled ore boxes are thrown around too, blocked the way, you cannot pull a single item for 3 seconds when the pressure gets worse, it's impossible to get through double firedoors as it takes time to open one of them, then it closes on you while you try to open the next one.

**Comment by Sniblet** [comment]
Monster the hero cures Yogs of stunmeta by replacing it with spacemeta
All antags just have to wear space protection and poke holes to cripple the station completely
This would be an interesting change if every round was extended, but they're not, and now if you just break a few walls all atmosphere in like a third of the station is gone, firelocks be damned
God help you and call the shuttle if engineering is bad or if distro is sabotaged. I can't wait to see what the next nightmare we spawn does once it gets its hands on an RCD

**Comment by LoliconSlayer** [comment]
**Who the fuck asked for this and how soon can I lynch them?**

**Comment by Bibby0110** [comment]
Please refrain from having an opinion on this game changing PR without playing the game with it on.

**Comment by ghost** [comment]
The test was a great success and Monster coded well. I think most of you should calm down and consider that if new mechanics are introduced, counterplay is also introduced. Now that the destructive side of an environmental system has been added, it is time to add in the defenses against it. The following are proposed changes I've spent some time mulling over and should address 95% of the issues raised in the discord and in ooc. 

-Firelocks are no longer regular airlocks attached to a fire alarm. Instead, they now have the ability to instantly dispense short-duration holofirelocks before they close. This safety precaution heavily limits the extent of spacing to whatever area is bounded by the firelocks instead of every bit of connected main hallway. Furthermore, the holofirelocks should be active for at least 5 seconds to give trapped people a good chance to get out without needing a crowbar. Alternatively, spawn a crowbar in the roundstart box, much easier than the FTL13 short duration spacesuit.

The holofirelock also fits the theme of an advanced research space station. Modern ships use bulkheads and pressure-resistant compartmentalized doors in the event of flooding, in space future, these sorts of things will be commonplace and encouraged for safety

-Mappers should add more firelock bulkheads around the map. 

-Mappers should also make sure windows that have access to space are filled out so they don't have access to space and use the fake space flooring that clowns love to use so much so that shitlers don't space everyone to shit by breaking internal windows.

-Spriters should make it so that bolted open firelocks should have the red bolt lights go off even when they are open. Welded open firelocks should have obvious sprites too. This stops the AI from simply bolting all firelocks open without people noticing and spacing everyone, unless they turn off bolt lights which is incredibly time consuming. 

Using ships as an example again, they are designed with great fear of taking on water and have many, many precautions to defend against them. I say we should attempt to model decompression precautions on flooding precautions, which guard against most of everything except:

-Explosives detonating in compartments due to enemy artillery or bombs in wartime, this is mirrored by syndi bombs going off, or other sources of explosions. The defense against this is rapid pumping and bulkhead compartmentalization - we already have rapid pumping in the form of vents and proposed firelock changes being implemented would supply the later
-Reserve buoyancy in the citadel in case of too much flooding. We have this in atmosia already. I propose an increase in the reserve gas tanks to compensate and a map change for a better optimized roundstart atmosia to cope in case of a lack of atmos techs or CEs. 

-Additional changes that could be added would be instant deployment of holofirelocks upon detecting plasma to cuck AIs, or at least make it a lot harder to plasmaflood. 

TL;DR - Overall, it makes sense that an advanced starship/station has lots of defenses against explosive decompressions. Firelocks should be buffed next, and much more massively.

**Comment by Bibby0110** [comment]
TL;DR 
Overall; It makes sense for a game to not worry about being blown out into space because window break

**Comment by MegaEmpirical** [comment]
It's an atmos sim, it makes sense.

**Comment by MegaEmpirical** [comment]
We could also consider increasing the strength of reinforced glass. Fuck the greytide.

**Comment by Bibby0110** [comment]
It's a game. It hasn't been an atmos sim for a while

**Comment by MegaEmpirical** [comment]
I didn't say it wasn't a game. Sims are an entire genre of games. I love Banished, a shame about the mid to late game though.

**Comment by Sniblet** [comment]
Since when is SS13 an atmos sim? Since a decade ago? 17 years? It's not an atmos sim anymore. That's exactly what this PR seems to think it is. It's absolutely not anymore, nobody even plays atmos. Your firelocks and real-life comparisons don't mean a thing when you're dealing with nukies or juggernauts or a blob or xeno acid or standard holoparasites or anyone with a Gygax Minecraft stripmining through walls to make a third of the station completely uninhabitable instantaneously. You can cheese the Wizard academy like this by instantly putting all of its inhabitants in vacuum. You can wipe departments by discreetly sabotaging or removing firelocks and then placing >one hole. And that's small-time compared to the worst you could do with this. Soon we'll have a revenant bust the vault windows while a small crowd is nearby and they'll all get electric-chainstunned and spaced to death faster than they can scream.

Where's the _good_ in this? The bad is that it completely changes the game and demands sweeping changes to all maps, roles, gamemodes, and players' attitudes without any warning or asking anyone for permission.
As for the good... I guess it's more realistic? :taptap: :rip: Okay, here are some suggestions so we can make our server as realistic as it's quite clearly aiming to be.
Port Bay med. Enforce the second law of thermodynamics. Get rid of chem dispensers, copy Charlie Station's chem over to the main station. Make space dust and meteors a constant phenomenon. Put finite batteries in all flashlights. Make space highly radioactive and remove temperature from it. Remove genetics, xenobiology, and mutations completely. Fully rework radiation. Make lava burn you just from proximity. Make all jobs so complicated that you actually need to study them extensively like a real doctor, or roboticist, or lawyer.
Or how about we don't?

I remember just yesterday OOC was joking about terrible PRs to add to the game, and I thought of "Hollywood decompression." It's real now. You did it. You absolute buffoon.

**Comment by ghost** [comment]
A buff for hull plating to have massively increased health so that only really strong bombs or an RCD will remove them would also heavily help. Making the RCD floor removal take 10 seconds and have sparks come out is also a good idea.

**Comment by monster860** [comment]
Yogsmeta may look like its missing mapmerge, but somehow this added just enough keys to require having 3-character keys instead of 2-character ones. Go figure.

**Comment by MegaEmpirical** [comment]
How will this affect flame throwers?

Grenades and smoke machine too I guess. Bombs?

**Comment by Readystorm546** [comment]
> Since when is SS13 an atmos sim? Since a decade ago? 17 years? It's not an atmos sim anymore. That's exactly what this PR seems to think it is. It's absolutely not anymore, nobody even plays atmos. <

No one even plays atmos lol. bruh calm down there is a TESTMERGE for a reason. If we just sissied around all the time nothing would get done and if this makes  you not wanna play im sorry but go ahead

**Comment by MegaEmpirical** [comment]
Several people play atmos and it's one of the best jobs on the station. Not sure where you guys are getting this information.

But yeah, it's a test. One, to see how to improve it and two, to see if it's viable.

**Comment by Bibby0110** [comment]
We tested flamethrowers and it makes them be slow since atmos is slower too
It almost looks like lag when you use a flamethrower

**Comment by boodaliboo** [comment]
cuurent implimentation i dont like it but if its change ill rethink that

**Comment by Darkstick** [comment]
The way this currently works is a complete game changer, and we would need to rework practically every aspect of the game around it. I'm not saying it's impossible but it's a ton of work and will modify the hell out of the meta. 

All that for what..? To simulate that one scene in sci-fi movies where someone gets sucked out of a window into space? To make one single job, atmos, more interesting?

I feel like this isn't worth the massive issues this will cause for the sake of a change doesn't seem to do anything positive gameplay-wise.

**Comment by Mqiib** [comment]
I really dislike this. It causes a bunch of weird differences in tiles where two adjacent tiles will have significant differences in pressure and temperature and remain that way for longer than is reasonable.

**Comment by monster860** [comment]
It doesnt cause those differences, they've always been there, it just doesn't fix the differences.

**Comment by Potototatoh** [comment]
improve dont remove

fag

**Comment by Alexandria404** [comment]
very nice
if it is too quick maybe make it slightly less fast

**Comment by Ratvar** [comment]
Add firelocks to mining shuttle areas, escape shuttle areas, different bar areas, cargo shuttle... and add ways to build new window shutters on top of grilles
I myself believe that walls, windows are too weak, and bombs/fire are too strong with new changes, and monstermos shouldn't be added without buffing first and nerfing second

**Comment by BeatoWalker** [comment]
This update is very good and it gives a terrific example of why you aren't supposed to go diving in space. There is some current change that I saw when I play the game during the test and added changes to the previous one. which is :

When Introduced : 
1. Airlock
you can get flushed real quick if someone forgot to cycle the airlock (In arrival, engineering (Tesla or singular engine built), sollar, and escape. it makes all the stuff get thrown all over and atmosphere unbreathable.

2. Large Section room
A breach is very dangerous especially on the station hallway where there was a large unprotected open space (example: north, east, and west primary hallway)

3. Window
The window that leading to space (example: near the Ai Core, near the Bridge, The arrival bay, the escape area) is very hazardous to the atmos breach

4. Scrubber and Vent
They added that it would be faster to scrub and Vent the air ( i think? not familiar and not expert with the code).

5. Movement of Air
The air changes and moves fast now I think, I'm not an expert in this and not familiar with it (example plasma flood is very dangerous now and can spread fast)

6. Spaced man -  Yeet
You can get thrown out to space if you are near when the breach occurred at.

7. Atmos holoair tool
I don't think it works with the air system when I test it and I still get sucked and thrown away when the breach occurs.

8. Shuttle
Someone can break the door, wall, or glass and will kill most of the player if there were a breach occurred there because of the lack of Air.

9. Magboots
I only know a few info about this and I heard it doesn't help much from being thrown away because of the depressurization event, because it's magnetic.

10.RCD
We only can build Airlock only, but not including the Shutter or The small Air door.

Added Changes
1. Airlock
there are two sets of air door added on the small airlock which prevents someone gets thrown out into space and prevent the room from getting unhabitable. You need a crowbar to open those Air door to get to the other side.

2. Large Section Room
There's now some small airlock that divides the hallway into each section of its own ( example the centre hallway are divided into 7 sections) and if there is a breach occur, the other section can be safely secured and habitable atmosphere

3. Window
I still don't know about the change regarding this. There are many shutters on every window (example: at the escape bay, Near Air Core, Near Vault, Near Arrivals, maintenance window and the Bridge)

4. Scrubber and Vent
I'm not familiar and not expert with this, so I won't comment on this. I heard from the chat that the scrubber and Vent system are much faster than the previous version.

5. Movement of Air
Not expert and familiar with this, soo I'm not commenting (and need help with this )

6. Spaceman - Yeet
You ran the f**k out from there if there's one or you will get spaced, what do you expect? (sarcasm)

7. Atmos Holo tools
I still don't know about this.

8. Shuttle
There are two sets of small airlock on the shuttle airlock which prevents it from the space breach. ( i think that's it?)

9. Magboots
No info is known yet.

10. RCD
We only can build Airlock only, but not including the Shutter or The small Airlock. (still the same, it would be a handful if we can build it too

11. Weapon and guns
currently, if you shoot a gun through a small airlock, it will pass through them

I test it when the test occurred and overall this is an amazing chance for the station crew experience. if there were more changes added, the game can be more playable and be more fun by those acceptable /viable changes. This might be a "major change" if accepted because of the impact on other things in the game 

Overall experience :
Overall got spaced twice and get noted from the admin, The risk and the result are Worth It. I'm gonna get noted again. 10/10

PS = Coder and mapper don't forget to sleep and drink more water because of these changes

**Comment by HomocidlePancakes** [comment]
Came out of my hole to say "Based". Based.
